File: leetcode/173.py
class BSTIterator:
    def __init__(self, root):
        self.stack = list()
        self.pushAll(root)
    def hasNext(self):
        return self.stack
    def next(self):
        tmpNode = self.stack.pop()
        self.pushAll(tmpNode.right)
        return tmpNode.val     
    def pushAll(self, node):
        while node is not None:
            self.stack.append(node)
            node = node.left

File: leetcode/173.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
private:
    stack<TreeNode*> s;
public:
    BSTIterator(TreeNode *root) {
        while (s.size())    s.pop();
        left(root);
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        if (s.size())  return 1;
        return 0;
    }

    /** @return the next smallest number */
    int next() {
        TreeNode* t = s.top();
        s.pop();
        if (t -> right) left(t -> right);
        return t -> val;
    }
    void left(TreeNode* rt) {
        TreeNode* t = rt;
        while(t) {
            s.push(t);
            t = t -> left;
        }
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */